mmcv/ops/points_sampler.py

```diff
@@ -0,0 +1,188 @@
+        self.fp16_enabled = False
+
+    @force_fp32()
+    def forward(self, points_xyz, features):
+        """forward.
+
+        Args:
+            points_xyz (Tensor): (B, N, 3) xyz coordinates of the features.
+            features (Tensor): (B, C, N) Descriptors of the features.
+
+        Return：
+            Tensor: (B, npoint, sample_num) Indices of sampled points.
+        """
+        indices = []
+        last_fps_end_index = 0
+
+        for fps_sample_range, sampler, npoint in zip(
+                self.fps_sample_range_list, self.samplers, self.num_point):
+            assert fps_sample_range < points_xyz.shape[1]
+
+            if fps_sample_range == -1:
+                sample_points_xyz = points_xyz[:, last_fps_end_index:]
+                sample_features = features[:, :, last_fps_end_index:] if \
+                    features is not None else None
+            else:
+                sample_points_xyz = \
+                    points_xyz[:, last_fps_end_index:fps_sample_range]
+                sample_features = \
+                    features[:, :, last_fps_end_index:fps_sample_range] if \
+                    features is not None else None
+
+            fps_idx = sampler(sample_points_xyz.contiguous(), sample_features,
+                              npoint)
+
+            indices.append(fps_idx + last_fps_end_index)
+            last_fps_end_index += fps_sample_range
+        indices = torch.cat(indices, dim=1)
+
+        return indices
+
+
+class DFPS_Sampler(nn.Module):
+    """DFPS_Sampling.
+
+    Using Euclidean distances of points for FPS.
+    """
+
+    def __init__(self):
+        super(DFPS_Sampler, self).__init__()
+
+    def forward(self, points, features, npoint):
+        """Sampling points with D-FPS."""
+        fps_idx = furthest_point_sample(points.contiguous(), npoint)
+        return fps_idx
+
+
+class FFPS_Sampler(nn.Module):
+    """FFPS_Sampler.
+
+    Using feature distances for FPS.
+    """
+
+    def __init__(self):
+        super(FFPS_Sampler, self).__init__()
+
+    def forward(self, points, features, npoint):
+        """Sampling points with F-FPS."""
+        assert features is not None, \
+            'feature input to FFPS_Sampler should not be None'
+        features_for_fps = torch.cat([points, features.transpose(1, 2)], dim=2)
+        features_dist = calc_square_dist(
+            features_for_fps, features_for_fps, norm=False)
+        fps_idx = furthest_point_sample_with_dist(features_dist, npoint)
+        return fps_idx
+
+
+class FS_Sampler(nn.Module):
+    """FS_Sampling.
+
+    Using F-FPS and D-FPS simultaneously.
+    """
+
+    def __init__(self):
+        super(FS_Sampler, self).__init__()
+
+    def forward(self, points, features, npoint):
+        """Sampling points with FS_Sampling."""
+        assert features is not None, \
+            'feature input to FS_Sampler should not be None'
+        features_for_fps = torch.cat([points, features.transpose(1, 2)], dim=2)
```

Comment:
We can create DFPS_Sampler and FFPS_Sampler instance and use their forward here.